drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+			dut_tcs[1] = data->mag_comp.dut_tcs.tcs_y;
+			dut_tcs[2] = data->mag_comp.dut_tcs.tcs_z;
+
+			/* Compensate raw magnetic data */
+			for (indx = 0; indx < 3; indx++) {
+				out_data[indx] = (out_data[indx] * (BMM350_MAG_COMP_COEFF_SCALING +
+								    dut_sensit_coef[indx])) /
+						 BMM350_MAG_COMP_COEFF_SCALING;
+				out_data[indx] = (out_data[indx] + dut_offset_coef[indx]);
+				out_data[indx] =
+					((out_data[indx] * BMM350_MAG_COMP_COEFF_SCALING) +
+					 (dut_tco[indx] * (out_data[3] - data->mag_comp.dut_t0))) /
+					BMM350_MAG_COMP_COEFF_SCALING;
+				out_data[indx] =
+					(out_data[indx] * BMM350_MAG_COMP_COEFF_SCALING) /
+					(BMM350_MAG_COMP_COEFF_SCALING +
+					 (dut_tcs[indx] * (out_data[3] - data->mag_comp.dut_t0)));
+			}
+
+			cr_ax_comp_x =
+				((((out_data[0] * BMM350_MAG_COMP_COEFF_SCALING) -
+				   (data->mag_comp.cross_axis.cross_x_y * out_data[1])) *
+				  BMM350_MAG_COMP_COEFF_SCALING) /
+				 ((BMM350_MAG_COMP_COEFF_SCALING * BMM350_MAG_COMP_COEFF_SCALING) -
+				  (data->mag_comp.cross_axis.cross_y_x *
+				   data->mag_comp.cross_axis.cross_x_y)));
+
+			cr_ax_comp_y =
+				((((out_data[1] * BMM350_MAG_COMP_COEFF_SCALING) -
+				   (data->mag_comp.cross_axis.cross_y_x * out_data[0])) *
+				  BMM350_MAG_COMP_COEFF_SCALING) /
+				 ((BMM350_MAG_COMP_COEFF_SCALING * BMM350_MAG_COMP_COEFF_SCALING) -
+				  (data->mag_comp.cross_axis.cross_y_x *
+				   data->mag_comp.cross_axis.cross_x_y)));
+
+			cr_ax_comp_z =
+				(out_data[2] +
+				 (((out_data[0] * ((data->mag_comp.cross_axis.cross_y_x *
+						    data->mag_comp.cross_axis.cross_z_y) -
+						   (data->mag_comp.cross_axis.cross_z_x *
+						    BMM350_MAG_COMP_COEFF_SCALING))) -
+				   (out_data[1] * ((data->mag_comp.cross_axis.cross_z_y *
+						    BMM350_MAG_COMP_COEFF_SCALING) -
+						   (data->mag_comp.cross_axis.cross_x_y *
+						    data->mag_comp.cross_axis.cross_z_x))))) /
+					 (((BMM350_MAG_COMP_COEFF_SCALING *
+					    BMM350_MAG_COMP_COEFF_SCALING) -
+					   data->mag_comp.cross_axis.cross_y_x *
+						   data->mag_comp.cross_axis.cross_x_y)));
+
+			out_data[0] = (int32_t)cr_ax_comp_x;
+			out_data[1] = (int32_t)cr_ax_comp_y;
+			out_data[2] = (int32_t)cr_ax_comp_z;
+		}
+		LOG_DBG("mag data %d %d %d\n", out_data[0], out_data[1], out_data[2]);
+
+		if (rslt == BMM350_OK) {
+			if ((data->axis_en & BMM350_EN_X_MSK) == BMM350_DISABLE) {
+				mag_temp_data->x = BMM350_DISABLE;
+			} else {
+				mag_temp_data->x = out_data[0];
+			}
+
+			if ((data->axis_en & BMM350_EN_Y_MSK) == BMM350_DISABLE) {
+				mag_temp_data->y = BMM350_DISABLE;
+			} else {
+				mag_temp_data->y = out_data[1];
+			}
+
+			if ((data->axis_en & BMM350_EN_Z_MSK) == BMM350_DISABLE) {
+				mag_temp_data->z = BMM350_DISABLE;
+			} else {
+				mag_temp_data->z = out_data[2];
+			}
+			mag_temp_data->temperature = out_data[3];
+		}
+	} else {
+		rslt = -EINVAL;
+	}
+	return rslt;
+}
+
+static int bmm350_sample_fetch(const struct device *dev, enum sensor_channel chan)
```

Comment:
What happens if the sensor is suspended through PM API? We don't have a solid answer for this today with the fetch+get API other than the application doing the power management reference counting on the sensors. But if the device *is* suspended and this is called, what happens?

Reply:
No, but this might be surprising behavior to a caller, maybe worth a warning log

Reply:
I think this behaves as it should with the PM model.

Reply:
It doesn't make sense to me to clutter the driver with this. Other drivers will also just try to fetch, no matter if the power mode allows for it.

Reply:
So will the registers continue to hold a reading from some past moment in time then or be all zeros?